pkg/docker/docker.go

```diff
@@ -608,49 +601,54 @@ func (d *stiDocker) RunContainer(opts RunContainerOptions) error {
 		if err := d.RemoveContainer(container.ID); err != nil {
 			glog.Warningf("Failed to remove container %q: %v", containerName, err)
 		} else {
 			glog.V(4).Infof("Removed container %q", containerName)
 		}
-	}()
-
-	// Attach to the container.
-	glog.V(2).Infof("Attaching to container %q ...", containerName)
-	attachOpts := opts.asDockerAttachToContainerOptions()
-	attachOpts.Container = container.ID
-	if _, err = d.client.AttachToContainerNonBlocking(attachOpts); err != nil {
-		glog.Errorf("Unable to attach to container %q with options %+v: %v", containerName, attachOpts, err)
-		return err
-	}
-
-	// Start the container.
-	glog.V(2).Infof("Starting container %q ...", containerName)
-	if err := util.TimeoutAfter(DefaultDockerTimeout, "timeout after waiting %v for Docker to start container", func() error {
-		return d.client.StartContainer(container.ID, nil)
-	}); err != nil {
-		return err
 	}
+	dumpStack := func(signal os.Signal) {
+		if signal == syscall.SIGQUIT {
+			buf := make([]byte, 1<<16)
+			runtime.Stack(buf, true)
+			fmt.Printf("%s", buf)
+		}
+		os.Exit(1)
+	}
+	return util.NewInterruptHandler(dumpStack, removeContainer).Run(func() error {
+		// Attach to the container.
+		glog.V(2).Infof("Attaching to container %q ...", containerName)
+		attachOpts := opts.asDockerAttachToContainerOptions()
+		attachOpts.Container = container.ID
+		if _, err = d.client.AttachToContainerNonBlocking(attachOpts); err != nil {
+			glog.Errorf("Unable to attach to container %q with options %+v: %v", containerName, attachOpts, err)
+			return err
+		}
 
-	// Run OnStart hook if defined.
-	if opts.OnStart != nil {
-		if err = opts.OnStart(container.ID); err != nil {
+		// Start the container.
+		glog.V(2).Infof("Starting container %q ...", containerName)
+		if err := util.TimeoutAfter(DefaultDockerTimeout, "timeout after waiting %v for Docker to start container", func() error {
+			return d.client.StartContainer(container.ID, nil)
+		}); err != nil {
 			return err
 		}
-	}
 
-	// We either block waiting for a user-originated SIGINT, or wait for the
-	// container to terminate. When TargetImage is true, we're dealing with
-	// an invocation of `s2i build ... --run` so this will, e.g., run a web
-	// server and block until the user interrupts it. The other case is seen
-	// when running the assemble script or other commands that are meant to
-	// terminate in a finite amount of time.
-	if opts.TargetImage {
-		runContainerDockerRun(container, d, image)
-	} else {
+		// Run OnStart hook if defined.
+		if opts.OnStart != nil {
+			if err = opts.OnStart(container.ID); err != nil {
+				return err
+			}
+		}
+
+		if opts.TargetImage {
+			// When TargetImage is true, we're dealing with an invocation of `s2i build ... --run`
+			// so this will, e.g., run a web server and block until the user interrupts it (or
+			// the container exits normally).  dump port/etc information for the user.
+			dumpContainerInfo(container, d, image)
```

Comment:
Looks like GitHub lost one of my comments here, not hidden, gone to /dev/null...

Summarizing, my realization/question was if this will not block, but rather return immediately, then we'll proceed and reach the `return` statement, terminating `s2i` and killing the running container way too early?